packages/config/src/class-config.php

```diff
@@ -0,0 +1,116 @@
+<?php
+/**
+ * The base Jetpack configuration class file.
+ *
+ * @package automattic/jetpack-config
+ */
+
+namespace Automattic\Jetpack;
+
+use Automattic\Jetpack\Sync\Actions as Sync_Actions;
+use Automattic\Jetpack\Sync\Main as Sync_Main;
+use Automattic\Jetpack\Plugin\Tracking as Plugin_Tracking;
+use Automattic\Jetpack\Terms_Of_Service;
+
+/**
+ * The configuration class.
+ */
+class Config {
+
+	/**
+	 * The Jetpack class instance.
+	 *
+	 * @var Jetpack
+	 */
+	protected $jetpack;
+
+	/**
+	 * The initial setting values.
+	 *
+	 * @var Array
+	 */
+	protected $config = array(
+		'sync'                => false,
+		'sync_woocommerce'    => false,
+		'sync_wp_super_cache' => false,
+		'tracking'            => false,
+		'tos'                 => false,
+	);
+
+	/**
+	 * Creates the configuration class instance, initalized with the Jetpack object.
+	 *
+	 * @param \Jetpack $jetpack the main object to initalize everything with.
+	 */
+	public function __construct( \Jetpack $jetpack ) {
```

Comment:
Having the need to access the Jetpack object to me defeats the purpose of having this as a separate package. In other words, in my head, if a package needs the Jetpack object, then it's not really independent enough to be a package. I expect that packages would be independent and I can drop them in my plugin without needing Jetpack.

Reply:
That's a very good point, thank you. How would you rather have packages receive things like the connection manager? I prefer this way of dependency management because it's easier to replace them with test mocks this way. But I agree, maybe it's better to somehow pass the connection manager object directly to the class that needs it. The problem is - how? We don't have an object factory, or a single initialization entry point. That's what I planned the config object to be.

Reply:
One way to handle this would be to instantiate package objects directly everywhere and not keep a single instance of them, like we sometimes do now. Do you think that's a sustainable approach? For example, like 
```
$manager = new \Automattic\Jetpack\Connection\Manager ();
$manager->verify_xmlrpc_signature();
```

Reply:
> One way to handle this would be to instantiate package objects directly everywhere and not keep a single instance of them, like we sometimes do now. Do you think that's a sustainable approach? 

Yeah, we've done that before, and I think it's not sustainable, because sometimes we'll add actions and filters on object instantiation, and we wouldn't want to duplicate those. It should be fine for stateless classes, like the `Status` one.

Reply:
> How would you rather have packages receive things like the connection manager? I prefer this way of dependency management because it's easier to replace them with test mocks this way. But I agree, maybe it's better to somehow pass the connection manager object directly to the class that needs it. The problem is - how? We don't have an object factory, or a single initialization entry point. That's what I planned the config object to be.

I think with all the complexity we have, it would definitely make sense to have a separate registry for those object instances that we don't want to duplicate. I'd suggest taking a look at the Registry design pattern - it allows us to exactly what we want here. And if we use that in a package to make it portable and reusable everywhere, and we update all usages to use that package instead of the affected instances directly, we should be good.

I realize registries are usually not recommended because of the global state they introduce, but using global state is exactly what we need to do here.

What do you think about that?

Reply:
Yes, the Registry pattern is something that I wanted to look into as well, but here we're trying to solve the problem of initialization order, so I guess we'll have to look into it separately. But that's a very interesting approach, thanks for that discussion!

Reply:
I agree it could be observed a separate problem, but I'll have to disagree with what this PR suggests - if we introduce dependencies to the main Jetpack class to packages in this PR, then it's a problem that should be solved before this PR. I'm pretty hesitant with the current state as it feels like a step back.

Reply:
> if we introduce dependencies to the main Jetpack class to packages in this PR, then it's a problem that should be solved before this PR
I agree here, this was very valuable feedback from you, hence I removed all references to the main Jetpack class from Sync and from Config too in a recent update. Now they are dependent only on other packages.

Reply:
Thanks Igor!